garage/tf/algos/reps.py

```diff
@@ -0,0 +1,442 @@
+                                      4)
+
+        if is_recurrent:
+            raise NotImplementedError
+
+        policy_dist_info_flat = self.policy.dist_info_sym(
+            i.flat.obs_var,
+            i.flat.policy_state_info_vars,
+            name="policy_dist_info_flat")
+
+        policy_dist_info_valid = filter_valids_dict(
+            policy_dist_info_flat,
+            i.flat.valid_var,
+            name="policy_dist_info_valid")
+
+        with tf.name_scope("bellman_error"):
+            delta_v = i.valid.reward_var + tf.tensordot(
+                i.feat_diff, i.param_v, 1)
+
+        with tf.name_scope("policy_loss"):
+            ll = pol_dist.log_likelihood_sym(i.valid.action_var,
+                                             policy_dist_info_valid)
+            loss = -tf.reduce_mean(ll * tf.exp(
+                delta_v / i.param_eta - tf.reduce_max(delta_v / i.param_eta)))
+
+            reg_params = self.policy.get_params(regularizable=True)
+            loss += self.l2_reg_loss * tf.reduce_sum(
+                [tf.reduce_mean(tf.square(param))
+                 for param in reg_params]) / len(reg_params)
+
+        with tf.name_scope("kl"):
+            kl = pol_dist.kl_sym(
+                i.valid.policy_old_dist_info_vars,
+                policy_dist_info_valid,
+            )
+            pol_mean_kl = tf.reduce_mean(kl)
+
+        with tf.name_scope("dual"):
+            dual_loss = i.param_eta * self.epsilon + i.param_eta * tf.log(
+                tf.reduce_mean(
+                    tf.exp(delta_v / i.param_eta -
+                           tf.reduce_max(delta_v / i.param_eta)))
+            ) + i.param_eta * tf.reduce_max(delta_v / i.param_eta)
+
+            dual_loss += self.l2_reg_dual * (
+                tf.square(i.param_eta) + tf.square(1 / i.param_eta))
+
+            dual_grad = tf.gradients(dual_loss, [i.param_eta, i.param_v])
+
+        self.f_dual = tensor_utils.compile_function(
+            flatten_inputs(self._dual_opt_inputs),
+            dual_loss,
+            log_name="f_dual")
+
+        self.f_dual_grad = tensor_utils.compile_function(
+            flatten_inputs(self._dual_opt_inputs),
+            dual_grad,
+            log_name="f_dual_grad")
+
+        self.f_policy_kl = tensor_utils.compile_function(
+            flatten_inputs(self._policy_opt_inputs),
+            pol_mean_kl,
+            log_name="f_policy_kl")
+
+        return loss
+
+    def _dual_opt_input_values(self, samples_data):
+        """Update dual func optimize input values based on samples data."""
+        policy_state_info_list = [
+            samples_data["agent_infos"][k]
+            for k in self.policy.state_info_keys
+        ]   # yapf: disable
+        policy_old_dist_info_list = [
+            samples_data["agent_infos"][k]
+            for k in self.policy.distribution.dist_info_keys
+        ]
+
+        self.feat_diff = self._features(samples_data)
```

Comment:
Features will be recalculated every time when dual_optimizer search for eta and v. This is not necessary.

Reply:
I remember I put it here for a reason. But I will check if it is unnecessary here. Thanks

Reply:
`feat_diff` is used to calculate the bellman error. This is required here.

Reply:
I've moved the line out of this function.